### DISCUSSION TITLE
Climate: on/off state should be independent from the HVAC mode

### DISCUSSION THREAD

**Discussion by Noltari** [home-assistant/architecture#998] [Entity Models]
Some modern climate systems have a master thermostat in which the HVAC mode can be changed and slave thermostats where the HVAC mode can't be changed (it will be forced to the one from its corresponding master thermostat):
- https://www.home-assistant.io/integrations/airzone
- https://www.home-assistant.io/integrations/airzone_cloud
- https://www.home-assistant.io/integrations/advantage_air

Master Zones allow:
- Changing the HVAC mode.
- Turning on/off.
- Changing the desired temperature.

Slave Zones allow:
- Turning on/off.
- Changing the desired temperature.

On Home Assistant climate entities we don't really support turning on a climate entity, because we consider any HVAC mode != Off as being On.
We should make the on/off status independent from the HVAC mode.
Most of the air conditioners have this behaviour: you can switch them on and off without changing its HVAC mode.

Related discussions:
- https://github.com/home-assistant/architecture/discussions/625
- https://github.com/home-assistant/architecture/discussions/982

**Comment by Noltari** [comment]
Moreover, some Air Conditioners such as Airzone have a "Stop Mode" (this is an additional HVAC mode), in which the compressor won't be supplied with electricity and the device will be set to the minimum power consumption.

**Reply by gjohansson-ST** [reply]
With the implementation of the new climate features for `turn_on` and `turn_off` isn't this "solved" by such?

**Reply by Noltari** [reply]
That would be the case if there weren't master and slave devices.
For Airzone devices, the HVAC mode shouldn't be changed when turning on or off a device, since other devices might be running and changing the HVAC mode from COOL/HEAT to STOP, would affect those other slave devices.
That's why we need an independent (switch like - ToggleEntity) parameter on Climate entities to control the on/off functionality without changing the HVAC mode.

**Reply by gjohansson-ST** [reply]
Not sure I understand this. If you implement the methods for `turn_on` and `turn_off` are you not then in control what they would do when called?

**Reply by OnFreund** [reply]
I don't know Airzone, but many A/Cs have independent controls for on/off and mode, and allow you to control the mode while off (thus determining what the next "On" state would be). There has been lots of feedback about this when Climate 1.0 was implemented, but unfortunately it wasn't taken into account, and the resulting design was basically optimizing for heat pumps, at the expense of A/Cs.

**Reply by BradleyFord** [reply]
Agree

The Daikin ACs have separate on/off and modes.

To work around this I need to use a HACS add-in that creates custom climate templates, and also addition code on how to blend together these into one type for HA.

Note: These are heat pump A/Cs

Daikin are one of the most significant players in the industry and often a standard setter

**Reply by OnFreund** [reply]
In this case you don't even need a standard setter - A/Cs have had separate knobs for mode and on/off for decades :)

**Reply by Noltari** [reply]
Sorry @gjohansson-ST, but I totally forgot to answer...

> Not sure I understand this. If you implement the methods for `turn_on` and `turn_off` are you not then in control what they would do when called?

Yes, but there's no ON/OFF button/switch at UI level for climate entities (thermostat cards) and I don't think that it will be possible to add it until proper independent on/off state support is added to climate entities, since you can only call `turn_on` and `turn_off`, but you don't have a binary status to show on the UI.

In my opinion, `OFF` should be removed from `HVACMode` and another binary attribute with on/off status should be added to `ClimateEntity`.
Therefore, climate devices which only support a single `HVACMode` should only be toggable on/off and the HVAC Mode shouldn't be modifiable, being always HEAT (or whatever that climate device supports).

**Reply by Noltari** [reply]
> Sorry @gjohansson-ST, but I totally forgot to answer...
> 
> > Not sure I understand this. If you implement the methods for `turn_on` and `turn_off` are you not then in control what they would do when called?
> 
> Yes, but there's no ON/OFF button/switch at UI level for climate entities (thermostat cards) and I don't think that it will be possible to add it until proper independent on/off state support is added to climate entities, since you can only call `turn_on` and `turn_off`, but you don't have a binary status to show on the UI.
> 
> In my opinion, `OFF` should be removed from `HVACMode` and another binary attribute with on/off status should be added to `ClimateEntity`. Therefore, climate devices which only support a single `HVACMode` should only be toggable on/off and the HVAC Mode shouldn't be modifiable, being always HEAT (or whatever that climate device supports).

@joostlek @gjohansson-ST in the end what I pretend to achieve is having a Toggle in the interface, like in the Light Entities:
![image](https://github.com/user-attachments/assets/b9a3e964-0ac1-40cd-82b6-78a159a8e51b)

And we could add some flags to the climate entity that would enable independently in the UI:
- Switch on/off.
- Changing HVAC mode.
Therefore, some climate entities would be able to do one of those two, or even both (most Daikin devices and Airzone master zones, for example).

**Comment by psychoboust** [comment]
Hi everyone, any news about this ?
I am new to Home Assistant, and was looking for a way to get the STOP MODE and also the ON/OFF button independant from the HVAC mode

**Comment by jbouwh** [comment]
I agree the climate entity should have an independent on/off state. I suggest we implement the `is_on` property.

property | type | default | description
-- | -- | -- | --
is_on | boolean | None | If the climate is currently on or off.

The climate entity already has turn_on and turn_off actions but it does not use `_attr_is_on`. The proposal is to add a `is_on` property to the `climate` entity platform. If the `_attr_is_on` attribute is set to `False` the state should always return `HVACMode.OFF.value` regardless of the current mode and action. It is up the the integration to decide if an independent on/off state is needed. IMO a climate can be `off` while the mode is `idle` or `auto`. The integration should be able to maintain the current state correctly.

PR to implement this: https://github.com/home-assistant/core/pull/134311

**Reply by gjohansson-ST** [reply]
Is that really the case?
I read the initial proposal to have a feature flag for allowing to set the mode rather than changing the state of the entity to on/off (`ToggleEntity`). Don't know what an additional property for `is_on` would make by itself. 

Anyhow, the proposal is lacking what is needed and how it's supposed to work.

**Reply by jbouwh** [reply]
You are right. Climate should not be a ToggleEntity, but it could implement the `is_on` property.
I'll open a PR to make a suggestion for implementation.

**Reply by jbouwh** [reply]
https://github.com/home-assistant/core/pull/134311

**Reply by gjohansson-ST** [reply]
Yes, but that would just add a property which in itself isn't used for anything.
I think what the base proposal was is to have a feature flag for setting mode so it's possible to toggle the entity on/off but not possible to set the mode (as in the base example, as I understand it, it's inherited from a master device).

Unless we think that changing the state from mode to an on/off I don't see the use of this property actually.

**Reply by jbouwh** [reply]
An issue I see is that devices have a power state and a mode state that are independent. I see this attribute to be useful in cases where these states are reported via 2 events like here:
https://github.com/home-assistant/core/issues/133556
But may be a `power_off_topic` and template is a better solution to override the mode.

**Reply by OnFreund** [reply]
> Climate should not be a ToggleEntity

Why not? I believe most HVAC systems are toggle-able, and google also [suggests the OnOff trait for A/Cs](https://developers.home.google.com/cloud-to-cloud/guides/acunit#onoff).

**Reply by OnFreund** [reply]
> Unless we think that changing the state from mode to an on/off I don't see the use of this property actually.

I think changing the state to on/off is the right approach. Mode is no different from temperature, which is controlled separately, and does not require the device to be on to control.

**Reply by jbouwh** [reply]
The only big objection is that changing the state will break about every automation. We could add a state attribute `is_on` whenever the `_attr_is_on` attribute is set to expose the on/off state. Integration could override the `is_on` property if that is needed.

**Reply by parautenbach** [reply]
I've been following this discussion for a while and have recently been impacted by what the proposal says: It would certainly be convenient to have a binary state. 

My small contribution here is that an attribute is probably not the best approach, since there's been a move against attributes to rather make separate sensors. A separate binary sensor would thus be more in line with the current architectural practice. 

This is in fact my current solution:

```
template:
  - binary_sensor:
      - name: Living Room AC
        state: "{{ has_value('climate.living_room_ac') and not is_state('climate.living_room_ac', 'off') }}"
```

I think it would be appropriate to keep the mode as the climate entity's state.

**Reply by OnFreund** [reply]
> The only big objection is that changing the state will break about every automation

That's indeed unfortunate. Since `hvac_mode` is identical to the state, I think that we can add an on/off property, add a deprecation notice so people start relying on `hvac_mode` rather than state, and eventually make state be on/off.
What's important though, is the ability to control those independently. I want to be able to change mode while the device is off, in the same way that I can change the temperature or swing state.

**Reply by parautenbach** [reply]
@OnFreund I guess I cannot force you to read what I said above, but please don't break the current state. It is not necessary. In the binary sensor domain, one would expect on/off as states. For a climate domain entity, climate-related values make sense. The current implementation is fine. You cannot (and shouldn't) have two domains that are effectively the same. There's already a binary sensor domain. That should be used. Integrations providing climate entities should (can) provide additional on/off entities as binary sensors.

**Reply by BradleyFord** [reply]
Let me provide a user story

I have multiple ACs controlled over BACNET, that come into HA via MQTT. I am using "Climate Template" to map each AC to a HA device. 

The challenge is that Daikin have " Mode" and "Power state" as independent data points, also the feedback on AC state (ie running) is separate as there is a delay between turning on and actually running. So I have needed to add some complex custom logic to map this. In particular splitting Mode/power state, and joining Mode/power state, note that the mapping and logic then needs to be different for reading and writing which makes it quite complicated.

The current approach we have is way to complicated for the average Joe/Jane, and even for people who are reasonably competent. 

Whatever the approach that is taken it should support independent mode and power state, rather than clubbing it together.

**Reply by OnFreund** [reply]
> I guess I cannot force you to read what I said above

There's no need for that tone. I did read your comment, but I was replying to a different one. I also happen to disagree with your point. I've seen a fair share of HVAC systems in my life and, with the exception of modern cars, almost all of them had at least 3 controls:
1. An on/off switch
2. A mode selector (empty in systems with only one mode)
3. An "intensity" selector. In modern systems this is usually a temperature selector, but in older systems it was common to have one dial that combined both mode and intensity, going from max cool on one end, to max heat on the other.

All 3 are integral parts of the system, in the same way that a fan is both either on or off, and has some intensity, or a light has both and on/off state and brightness or color, or a media player tracks both its on/off state as well as playing state. While HA is moving away from attributes for auxiliary information, I've not seen a similar move when it comes to the core of the relevant domain, such as the examples I gave above.

Also, I think you mean switch, not binary sensor...

**Reply by jbouwh** [reply]
I updated the proposal a bit. The new `is_on` attribute would be a new climate property. When `_attr_is_on` is set, `is_on` will follow the state of `_attr_is_on`, else will be `False` if the current action is `OFF` if that is available or of the mode is `OFF`.
The state will be `OFF` is `_attr_is_on` is `False` even if the mode is e.g. `AUTO`. It is up to the integration to set the best suitable mode and current action (if that is supported).

**Reply by frenck** [reply]
Why not just add a switch entity for that device?

**Reply by Noltari** [reply]
> Why not just add a switch entity for that device? 

Yes, that was exactly what I was proposing.

**Reply by gjohansson-ST** [reply]
> > Why not just add a switch entity for that device?
> 
> Yes, that was exactly what I was proposing.

I think you started this proposal @Noltari in the attempt to change the `ClimateEntity` somewhat, but now it would be fine to use a `SwithcEntity` instead?

**Reply by OnFreund** [reply]
> Why not just add a switch entity for that device?

For the same reason we don't use a number entity for the temperature - the on/off state is an integral part of the state, and having it as a separate entity complicates the interface. I believe that HVAC systems that aren't toggle-able are the exception, not the rule, yet the current design optimizes for them.
Similar entities, such as humidifiers, and even fans (where ironically many physical devices are designed like the climate entity, meaning having no "on" switch, but rather "off" and a few presets), are all toggleable. I think Climate should follow suit.

**Reply by Noltari** [reply]
> > > Why not just add a switch entity for that device?
> > 
> > 
> > Yes, that was exactly what I was proposing.
> 
> I think you started this proposal @Noltari in the attempt to change the `ClimateEntity` somewhat, but now it would be fine to use a `SwithcEntity` instead?

@gjohansson-ST sorry for that, I misunderstood this and I thought the proposal was to make ClimateEntity a ToggleEntity, which I confused with the proposed SwitchEntity.
In any case I was just trying to make some progress since the discussion has been open for quite a long time...

**Comment by campinge** [comment]
Any way to push this? I'm also writing some logic currently and it requires a ton of additional logic (turning off all zones and then always making sure to turn them back on again when they are used) to actually force the unit to stop. Very unsatisfying :(

**Reply by mateuszdrab** [reply]
In my case my modbus controlled AC can be controlled fine from HA, it controls modes and can turn the AC on and off. However, if the AC is turned off using the remote, HA can't turn it on unless I manually send a turn on command via modbus. This makes me think the climate entity doesn't send an on command and instead just sets the mode, which the modbus controller doesn't properly translate into turning the unit on. I work around it by just never turning the AC off using the remote, but it is a bit weird.
